core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -417,16 +418,19 @@ private[spark] object Utils extends Logging {
       input: FileChannel,
       output: WritableByteChannel,
       startPosition: Long,
-      bytesToCopy: Long): Unit = {
+      bytesToCopy: Long,
+      numTransferToCalls: Int): Unit = {
     val outputInitialState = output match {
       case outputFileChannel: FileChannel =>
         Some((outputFileChannel.position(), outputFileChannel))
       case _ => None
     }
     var count = 0L
+    var num = 0
     // In case transferTo method transferred less data than we have required.
-    while (count < bytesToCopy) {
+    while (count < bytesToCopy && num < numTransferToCalls) {
```

Comment:
I mean, wouldn't this potentially mean you don't copy all the data?
This seems very hacky. Is it that some other setting means the transferTo calls barely transfer anything?
This doesn't seem like a Spark-level issue.

Reply:
Yes, in a situation, Spark cannot copy all the data anymore. I agree it is quite hacky, but simply we don't want Spark to hang indefinitely and make it fail fast.

This seems is not a Spark problem, but in a very big and elastic cluster, the chance of meeting such problem may not be rare, and hard to detect in the system level. It would be better to have a way to handle this in Spark side.